neqo-http3/src/recv_message.rs

```diff
@@ -132,32 +135,47 @@ impl RecvMessage {
         Ok(())
     }
 
-    fn add_headers(&mut self, headers: Option<Vec<Header>>, fin: bool, decoder: &mut QPackDecoder) {
-        if fin {
-            self.conn_events.header_ready(self.stream_id, headers, true);
-            self.set_closed(decoder);
-        } else {
-            self.conn_events
-                .header_ready(self.stream_id, headers, false);
-            self.state = RecvMessageState::WaitingForData {
-                frame_reader: HFrameReader::new(),
-            };
+    fn add_headers(
+        &mut self,
+        headers: Vec<Header>,
+        fin: bool,
+        decoder: &mut QPackDecoder,
+    ) -> Res<()> {
+        match (fin, self.is_interim(&headers)?) {
+            (true, true) => Err(Error::HttpGeneralProtocolStream),
+            (false, true) => {
+                self.conn_events
+                    .header_ready(self.stream_id, headers, true, fin);
+                self.state = RecvMessageState::WaitingForResponseHeaders {
+                    frame_reader: HFrameReader::new(),
+                };
```

Comment:
Is this just a case of needing to reset the value of frame_reader?  Can you comment to that effect?